src/win/util.c

```diff
@@ -1658,16 +1662,56 @@ int uv_os_uname(uv_utsname_t* buffer) {
   }
 
   /* Populate the version field. */
-  if (WideCharToMultiByte(CP_UTF8,
-                          0,
-                          os_info.szCSDVersion,
-                          -1,
-                          buffer->version,
-                          sizeof(buffer->version),
-                          NULL,
-                          NULL) == 0) {
-    r = uv_translate_sys_error(GetLastError());
-    goto error;
+  version_size = 0;
+  r = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                    L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion",
+                    0,
+                    KEY_QUERY_VALUE,
+                    &registry_key);
+
+  if (r == ERROR_SUCCESS) {
+    product_name_w_size = sizeof(product_name_w);
+    r = RegQueryValueExW(registry_key,
+                        L"ProductName",
+                        NULL,
+                        NULL,
+                        (BYTE*) &product_name_w,
+                        &product_name_w_size);
+    RegCloseKey(registry_key);
+
+    if (r == ERROR_SUCCESS) {
+      product_name_w[product_name_w_size / sizeof(WCHAR)] = L'\0';
```

Comment:
Doesn't this write one past the end when `product_name_w_size == sizeof(product_name_w)`?

Reply:
Sigh. Yes. So, this is just to ensure that the registry value is null terminated in case it was tampered with. Normally it shouldn't be necessary, and there is a good chance the value is bogus if it was tampered with. It's probably good enough to just set `product_name_w[ARRAY_SIZE(product_name_w) - 1] = L'\0'` and be done with it.

Reply:
You could use the "newer" [`RegGetValueW`](https://docs.microsoft.com/en-us/windows/desktop/api/winreg/nf-winreg-reggetvaluea) since it's within our minimal `_WIN32_WINNT = 0x0600`. It will always return a nul terminated string.
```
    product_name_w_size = sizeof(product_name_w);
    r = RegGetValueA(registry_key,
                    NULL,
                    "ProductName",
                    RRF_RT_REG_EXPAND_SZ | RRF_RT_REG_MULTI_SZ | RRF_RT_REG_SZ,
                    0,
                    (PVOID)product_name_w,
                    &product_name_w_size);
    RegCloseKey(registry_key);
```
so `product_name_w` is guaranteed to be nul terminated iff r == 0

Reply:
Thanks @refack! I'll try that approach - it seems much better.

Reply:
@refack check out https://ci.nodejs.org/job/libuv-test-commit-windows/1323/. It seems to work on all of the Windows CI hosts except the win 2008 host.